src/libraries/System.Runtime.Extensions/tests/System/Convert.FromHexString.cs

```diff
@@ -34,93 +35,125 @@ public static void CompleteValueRange()
+            AssertExtensions.SequenceEqual(expected, tryResult);
+
         }
 
         [Fact]
         public static void InvalidInputString_Null()
         {
             AssertExtensions.Throws<ArgumentNullException>("s", () => Convert.FromHexString(null));
+            AssertExtensions.Throws<ArgumentNullException>("source", () => Convert.FromHexString(null, default, out _, out _));
         }
 
-        [Fact]
-        public static void InvalidInputString_HalfByte()
+        [Theory]
+        [InlineData("01-02-FD-FE-FF")]
+        [InlineData("00 01 02FD FE FF")]
+        [InlineData("000102FDFEFF  ")]
+        [InlineData("  000102FDFEFF")]
+        [InlineData("\u200B 000102FDFEFF")]
+        [InlineData("0\u0308")]
+        [InlineData("0x")]
+        [InlineData("x0")]
+        public static void InvalidInputString_FormatException_Or_FalseResult(string invalidInput)
         {
-            Assert.Throws<FormatException>(() => Convert.FromHexString("ABC"));
-        }
+            Assert.Throws<FormatException>(() => Convert.FromHexString(invalidInput));
 
-        [Fact]
-        public static void InvalidInputString_BadFirstCharacter()
-        {
-            Assert.Throws<FormatException>(() => Convert.FromHexString("x0"));
+            Span<byte> buffer = stackalloc byte[invalidInput.Length / 2];
+            Assert.Equal(OperationStatus.InvalidData, Convert.FromHexString(invalidInput.AsSpan(), buffer, out _, out _));
         }
 
         [Fact]
-        public static void InvalidInputString_BadSecondCharacter()
+        public static void ZeroLength()
         {
-            Assert.Throws<FormatException>(() => Convert.FromHexString("0x"));
-        }
+            Assert.Same(Array.Empty<byte>(), Convert.FromHexString(string.Empty));
 
-        [Fact]
-        public static void InvalidInputString_NonAsciiCharacter()
-        {
-            Assert.Throws<FormatException>(() => Convert.FromHexString("0\u0308"));
-        }
+            OperationStatus convertResult = Convert.FromHexString(string.Empty, Span<byte>.Empty, out int consumed, out int written);
 
-        [Fact]
-        public static void InvalidInputString_ZeroWidthSpace()
-        {
-            Assert.Throws<FormatException>(() => Convert.FromHexString("\u200B 000102FDFEFF"));
+            Assert.Equal(OperationStatus.Done, convertResult);
+            Assert.Equal(0, written);
+            Assert.Equal(0, consumed);
         }
 
         [Fact]
-        public static void InvalidInputString_LeadingWhiteSpace()
+        public static void ToHexFromHexRoundtrip()
         {
-            Assert.Throws<FormatException>(() => Convert.FromHexString("  000102FDFEFF"));
-        }
+            const int loopCount = 50;
+            Span<char> buffer = stackalloc char[loopCount * 2];
+            for (int i = 1; i < loopCount; i++)
+            {
+                byte[] data = Security.Cryptography.RandomNumberGenerator.GetBytes(i);
```

Comment:
We usually use a seeded Random so the tests are reproducible.

Reply:
I agree with Dan, but since the existing tests in this type were already using `Security.Cryptography.RandomNumberGenerator` it's acceptable to keep it.